DESIGN.md

```diff
@@ -0,0 +1,225 @@
+controller that listens to netlink events and calls methods on the model and
+view instances in response to, say, a NIC gaining an address.
+
+The views display the model and call methods on the controller to make changes.
+
+Obviously for most screens there is a triple of a model class, controller class
+and a view class for the initial view, but this isn't always true -- some
+controllers don't have a corresponding model class.
+
+### Doing things in the background
+
+If the UI does not block, as promised above, then there needs to be a way of
+running things in the background.  The main way this is done is the `run_in_bg`
+function, which takes two functions: one that takes no arguments and is called
+in a background thread and a callback that takes one argument, and is called in
+the main/UI thread with a `concurrent.futures.Future` representing the result
+of calling the first function. This is a fairly clunky interface that can lead
+to some slightly baffling control flow but, well, it works. I've tried a few
+abstractions that make things clearer -- `subiquity.tasksequence.TaskSequence`
+being the older one that doesn't quite cover all interesting cases and
+`subiquity.controllers.installprogress.StateMachine` being the newer one that I
+hope can be generalized to something useful in many places.
+
+Currently subiquity just uses urwid's default event loop (`SelectEventLoop`).
+Switching to `AsyncioEventLoop` would probably be an improvement and might let
+us not use a background thread for absolutely every action we want to run in
+the background (there's no real reason to tie up a thread waiting for a
+subprocess to exit, for example).
+
+Only touch the UI from the main thread.
+
+### Terminal things
+
+Subiquity is mostly developed in a graphical terminal emulator like
+gnome-terminal, but ultimately runs for real in a linux tty or over a serial
+line.
+
+The main limitation of the linux tty is that it only supports a font with 256
+characters (and that's if you use the mode that supports only 8 colors, which
+is what subiquity does).  Subiquity comes with its own console font (see the
+`font/` subdirectory) that uses different glyphs for arrows and has a check
+mark character. gnome-terminal supports utf-8 of course, so that just works
+during development -- one just has to be a bit careful when using non-ascii
+characters. There are still plenty of characters in the standard font subiquity
+does not use, so we can add support for at least a dozen or so more glyphs if
+there's a need.
+
+`subiquity.palette` defines the 8 RGB colors and a bunch of named "styles" in
+terms of foreground and background colors.  `subiquitycore.core` contains some
+rather hair-raising code for mangling these definitions so that using these
+style names in urwid comes out in the right color both in gnome-terminal (using
+ISO-8613-3 color codes) and in the linux tty (using the PIO_CMAP ioctl).
+
+### Testing
```

Comment:
URLs to the platform qa test results maybe?

Reply:
Yeah added, along with the VPN related disclaimer.